src/lib.rs

```diff
@@ -917,6 +939,32 @@ impl<T, L, R> Either<(L, T), (R, T)> {
     }
 }
 
+/// `Either<L, R>` is an iterator if both `L` and `R` are iterators.
+impl<L, R> Either<L, R>
+where
+    L: IntoIterator,
+    R: IntoIterator,
+{
+    /// Factors an `Either` of `Itorator`s to be an `Itorator` of `Either`s
+    /// ```
+    /// use either::*;
+    /// let left: Either<_, Vec<u8>> = Left(&["hello"]);
+    /// assert_eq!(left.factor_iter().next(), Some(Left(&"hello")));
+
+    /// let right: Either<&[&str], _> = Right(vec![0, 1]);
+    /// assert_eq!(right.factor_iter().collect::<Vec<_>>(), vec![Right(0), Right(1)]);
+    ///
+    /// ```
+    // TODO(MSRV): doc(alias) was stabilized in Rust 1.48
+    // #[doc(alias = "transpose")]
+    pub fn factor_iter(self) -> impl Iterator<Item = Either<L::Item, R::Item>> {
```

Comment:
I wish there was a better return type here so that e.g. the result would be a `DoubleEndedIterator` if both sides were, but map returns types are ugly even when they are not ineffable.

Reply:
Yeah... it can _mostly_ be spelled out like this:

```rust
Either<
    iter::Map<L::IntoIter, impl Fn(L::Item) -> Either<L::Item, R::Item>>,
    iter::Map<R::IntoIter, impl Fn(R::Item) -> Either<L::Item, R::Item>>,
>
```

... and that will be conditionally double-ended and such. Or we can add a bespoke iterator type with a cleaner type signature.

Reply:
I updated the type signature to your suggestion. It is a bit icky to expose so much of the implementation in the type signature as it would require a semver bump to change it, but I can't see a better option. 

I thought about adding a bespoke iterator, but I was reluctant to polute `either`'s otherwise minimalist name space, and it was going to be a whole pile of boilerplate to forward all the methods to the correct side to pick up any optimisations (e.g. for `count` or `last`).

Reply:
It's not a great sign when the type signature is much longer than the implementation. :)

That almost makes me want to scrap the idea, except maybe it is still valuable to have something that guides type inference to the same `Item = Either<L::Item, R::Item>`. If you write that `map_either` implementation on its own, the `Left` side has ambiguous `R`, and the `Right` has ambiguous `L`. But from a few experiments, it does seem like the compiler's type inference figures out how to unify those if you try to use it as an `Iterator`, where they must be the same.